### PR TITLE
Add peephole optimizations for CFG blocks.

### PR DESCRIPTION
**Pull request by GabiTulba**
The optimizations added are as follows:
1. remove_useless_mov which treats the case:
        
        mov r1, r2
        mov r2, r1

        In this case the second move can be deleted

2. fold_intop_imm which treats the case:

        op1 r, imm1
        op2 r, imm2

        We can rewrite this as:

        op1 r, (imm1 op2 imm2)

        Iff op1 and op2 are associative and the result (imm1 op2 imm2) does not overflow.

Added the following extra functionality for the `Flambda_backend.Doubly_linked_list` data structure:
* `next cell` -> get the next cell
* `hd_cell t` -> get a cell to the first element in the list
* `last_cell t` -> get a cell to the last element of the list
* `delete_curr/delete_after/delete_before cell` -> delete the current element, the one after or before a cell. asserts if there is no cell to delete (i.e. we try to delete before the first element or after the last)
* extra functionlity for inserting and getting back the newly inserted cell in a `Doubly_linked_list`

Added:
* Flags for enabling peephole optimizations (`-cfg-peephole-optimize`)
* A generic CSV data structure

This was benchmarked on the following libraries with the following results:
* Base:
  - remove_useless_mov hits: `4145`
  - fold_intop_arith hits: `4`
  - fold_intop_bitwise hits: `8`
  - total build time without optimizations (avg of 10 runs): `171.54 sec`
  - total build time with optimizations (avg of 10 runs): `171.84 sec`
  - total `.o` file size reduction from optimizations: `2288 bytes`

* Core:
  - remove_useless_mov hits: `4473`
  - fold_intop_arith hits: `9`
  - fold_intop_bitwise hits: `51`
  - total build time without optimizations (avg of 10 runs): `177.83 sec`
  - total build time with optimizations (avg of 10 runs): `177.43 sec`
  - total `.o` file size reduction from optimizations: `5968 bytes`

### CODE REVIEW

File: backend/.ocamlformat-enable [hunk 1]
```diff
@@ -6,11 +6,15 @@ checks.ml
 checks.mli
 checkmach.ml
 checkmach.mli
+csv_data.ml
+csv_data.ml
```
Comment (xclerc) [inline backend/.ocamlformat-enable:10]:
```suggestion
```

File: dune [hunk 2]
```diff
@@ -193,6 +195,10 @@
   linscan
   liveness
   mach
+  peephole_optimize
+  peephole_rules
+  peephole_utils
+  csv_data
```
Comment (xclerc) [inline dune:201]:
```suggestion
```

File: backend/asmgen.ml [hunk 3]
```diff
@@ -308,6 +308,14 @@ let compile_fundecl ~ppf_dump ~funcnames fd_cmm =
         ++ Cfg_with_infos.cfg_with_layout
         ++ Profile.record ~accumulate:true "cfg_validate_description" (Regalloc_validate.run cfg_description)
         ++ Profile.record ~accumulate:true "cfg_simplify" Regalloc_utils.simplify_cfg
+        ++ Profile.record ~accumulate:true "peephole_optimize_cfg" (fun (cfg: Cfg_with_layout.t) ->
+          (* CR-someday gtulbalecu: The peephole optimizations must not affect livenes, otherwise
```
Comment (xclerc) [inline backend/asmgen.ml:312]:
```suggestion
          (* CR-someday gtulbalecu: The peephole optimizations must not affect liveness, otherwise
```

File: backend/csv_data.ml [hunk 4]
```diff
@@ -0,0 +1,104 @@
+[@@@ocaml.warning "+a-4-30-40-41-42"]
```
Comment (xclerc) [inline backend/csv_data.ml:1]: (This module is not used in this pull
request, and should be removed I
think.)

File: backend/csv_data.mli [hunk 5]
```diff
@@ -0,0 +1,27 @@
+[@@@ocaml.warning "+a-4-30-40-41-42"]
```
Comment (xclerc) [inline backend/csv_data.mli:1]: (This module is not used in this pull
request, and should be removed I
think.)

File: backend/peephole/peephole_rules.ml [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+    else None
+  | _ -> None
+
+let handbuilt_rules cell =
+  match remove_useless_mov cell with
+  | None -> ( match fold_intop_imm cell with None -> None | res -> res)
+  | res -> res
+
+let handbuilt_rule_names =
+  ["remove_useless_mov"; "fold_intop_imm_bitwise"; "fold_intop_imm_arith"]
```
Comment (xclerc) [inline backend/peephole/peephole_rules.ml:182]:
```suggestion
```

File: backend/peephole/peephole_rules.mli [hunk 7]
```diff
@@ -0,0 +1,7 @@
+[@@@ocaml.warning "+a-29-40-41-42"]
+
+open! Peephole_utils
+
+val handbuilt_rules : rule
+
+val handbuilt_rule_names : string list
```
Comment (xclerc) [inline backend/peephole/peephole_rules.mli:7]:
```suggestion
```

File: driver/flambda_backend_args.ml [hunk 8]
```diff
@@ -51,6 +51,12 @@ let mk_regalloc_validate f =
 let mk_no_regalloc_validate f =
   "-no-regalloc-validate", Arg.Unit f, " Do not validate register allocation"
 
+let mk_cfg_peephole_optimize f = 
+  "-cfg-peephole-optimize", Arg.Unit f, " Apply peephole optimizations to CFG representation block bodies"
```
Comment (xclerc) [inline driver/flambda_backend_args.ml:55]:
```suggestion
  "-cfg-peephole-optimize", Arg.Unit f, " Apply peephole optimizations to CFG"
```

File: driver/flambda_backend_args.ml [hunk 9]
```diff
@@ -51,6 +51,12 @@ let mk_regalloc_validate f =
...
+
+let mk_no_cfg_peephole_optimize f = 
+  "-no-cfg-peephole-optimize", Arg.Unit f, " Do not apply peephole optimizations to CFG representation block bodies"
```
Comment (xclerc) [inline driver/flambda_backend_args.ml:58]:
```suggestion
  "-no-cfg-peephole-optimize", Arg.Unit f, " Do not apply peephole optimizations to CFG"
```

File: utils/doubly_linked_list.ml [hunk 10]
```diff
@@ -29,31 +29,41 @@ let insert_before cell value =
...
-      match new_node.prev with
+      (match new_node.prev with
       | Empty -> cell.t.first <- value_node
-      | Node node -> node.next <- value_node))
+      | Node node -> node.next <- value_node);
+      { node = value_node; t = cell.t })
 
-let insert_after cell value =
+let insert_before cell value =
+  let _new_cell = insert_and_return_before cell value in
```
Comment (xclerc) [inline utils/doubly_linked_list.ml:43]: (To be really useful, the binding
should include the type.)

File: utils/doubly_linked_list.ml [hunk 11]
```diff
@@ -29,31 +29,41 @@ let insert_before cell value =
...
+       new_node.prev <- cell.node;
+       cell_node.next <- value_node;
+       cell.t.length <- succ cell.t.length;
+       match new_node.next with
+       | Empty -> cell.t.last <- value_node
+       | Node node -> node.prev <- value_node);
+      { node = value_node; t = cell.t })
+
+let insert_after cell value =
+  let _new_cell = insert_and_return_after cell value in
```
Comment (xclerc) [inline utils/doubly_linked_list.ml:65]: (Same here.)

File: utils/doubly_linked_list.ml [hunk 12]
```diff
@@ -149,6 +172,33 @@ let remove t curr =
...
+  | Empty ->
+    (* internal invariant: cell's nodes are not empty *)
+    assert false
+  | Node cell_node ->
+      match cell_node.next with
+      | Empty ->
+        (* convention: cannot delete_after the first element in the list *)
+        assert false
+      | Node next_cell_node ->
+        delete_curr {node=cell_node.next; t=cell.t}
```
Comment (xclerc) [inline utils/doubly_linked_list.ml:200]:
```suggestion
        delete_curr { node=cell_node.next; t=cell.t }
```
(This function is not formatted like the one above.)

File: backend/peephole/peephole_utils.mli [hunk 13]
```diff
@@ -0,0 +1,24 @@
+[@@@ocaml.warning "+a-29-40-41-42"]
+
+module DLL = Flambda_backend_utils.Doubly_linked_list
+
+val are_equal_regs : Reg.t -> Reg.t -> bool
+
+val go_back_const : int
```
Comment (xclerc) [inline backend/peephole/peephole_utils.mli:7]: The semantics of most values can
be inferred from their names, but I
think `go_back_const` and `rule`
would deserve a docstring.

File: backend/peephole/peephole_utils.ml [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+  | Mach.Iand | Ior | Ixor | Ilsl | Ilsr | Iasr -> true
+  | _ -> false
+
+let bitwise_overflow_assert (imm1 : int) (imm2 : int) (op : int -> int -> int) =
+  let imm = op imm1 imm2 in
+  assert (imm <= 2147483647 && imm >= -2147483648)
+
+let no_32_bit_overflow imm1 imm2 op =
+  let imm = op imm1 imm2 in
+  -2147483648 <= imm && imm <= 2147483647
```
Comment (xclerc) [inline backend/peephole/peephole_utils.ml:46]: (I would use `Int32.{min,max}_int`,
do the comparison in only one
function, and mark these functions
`[@inline]`.)

File: backend/peephole/peephole_utils.ml [hunk 15]
```diff
@@ -0,0 +1,50 @@
+(* CR-soon gtulba-lecu for gtulba-lecu: drop the -4 flag *)
```
Comment (xclerc) [inline backend/peephole/peephole_utils.ml:1]: Yes, maybe drop `-4` here and
disable the warning only for
function `is_bitwise_op`.

File: backend/peephole/peephole_rules.ml [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+    bitwise_overflow_assert imm1 imm2 ( lor );
+    Some (Mach.Ior, imm1 lor imm2)
+  | Ixor, Ixor ->
+    bitwise_overflow_assert imm1 imm2 ( lxor );
+    Some (Mach.Ixor, imm1 lxor imm2)
+  (* For the following three cases we have the issue that in some situations,
+     one or both immediate values could be out of bounds, but the result might
+     be within bounds (e.g. imm1 = -4 and imm2 = 65, their sum being 61), in
+     that case we still do the folding. This should not happen at all since the
+     immediate values should always be within the bounds [0, Sys.int_size]. *)
```
Comment (xclerc) [inline backend/peephole/peephole_rules.ml:65]: The code and the comment do not
look in sync. Indeed, I think we
should (conservatively) not simplify
in the `-4`, `65` case but we do
below.

File: backend/peephole/peephole_rules.ml [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+    if Misc.no_overflow_add imm1 imm2 && no_32_bit_overflow imm1 imm2 ( + )
+    then Some (Mach.Iadd, imm1 + imm2)
+    else None
+  | Iadd, Isub ->
+    if imm1 >= imm2
+    then
+      if Misc.no_overflow_sub imm1 imm2 && no_32_bit_overflow imm1 imm2 ( - )
+      then Some (Mach.Iadd, imm1 - imm2)
+      else None
+    else if Misc.no_overflow_sub imm1 imm2 && no_32_bit_overflow imm2 imm1 ( - )
```
Comment (xclerc) [inline backend/peephole/peephole_rules.ml:91]: I think it should be `Misc.no_overflow_sub imm2 imm1`.

File: backend/peephole/peephole_rules.ml [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+    if Misc.no_overflow_add imm1 imm2 && no_32_bit_overflow imm1 imm2 ( + )
+    then Some (Mach.Isub, imm1 + imm2)
+    else None
+  | Isub, Iadd ->
+    if imm1 >= imm2
+    then
+      if Misc.no_overflow_sub imm1 imm2 && no_32_bit_overflow imm1 imm2 ( - )
+      then Some (Mach.Isub, imm1 - imm2)
+      else None
+    else if Misc.no_overflow_sub imm1 imm2 && no_32_bit_overflow imm2 imm1 ( - )
```
Comment (xclerc) [inline backend/peephole/peephole_rules.ml:104]: (Same here.)

**Review by xclerc — COMMENTED** [review]
Looks pretty good; most of the suggested
changes are about unused elements and
typos.

Did not put a comment on the diff, but I
think it would make sense to use the
same set of warnings in all files. With
one possible exception: the "fragile
pattern" warning, but maybe we can
disable it locally rather than at the
top of some files.

File: backend/peephole/peephole_rules.ml [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+          DLL.delete_curr snd;
+          Some ((prev_at_most go_back_const) new_cell)
+        | _ -> None)
+      | _ -> None
+    else None
+  | _ -> None
+
+let handbuilt_rules cell =
+  match remove_useless_mov cell with
+  | None -> ( match fold_intop_imm cell with None -> None | res -> res)
```
Comment (xclerc) [inline backend/peephole/peephole_rules.ml:178]: (The CI will probably fail because of `ocamlformat`.)

File: backend/peephole/peephole_utils.mli (hunk 13 above)
Comment (GabiTulba) [reply backend/peephole/peephole_utils.mli:7]: We agreed to only ad a docstring for `rule` as `go_back_const` will be deleted soon.

**Review by gretay-js — CHANGES_REQUESTED** [review]
Can you add some tests, including overflow tests?

Additionally, there should be a check to guarantee termination of rewriting (e.g., if one pattern removes instruction and another adds it back). Currently, patterns only remove instructions so this is implicit, but may worth a comment somewhere.

Separate architectural constraints into a helper function, and leave refactoring it into target specific folder to another PR.

File: backend/peephole/peephole_utils.mli [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+
+val is_bitwise_op : Mach.integer_operation -> bool
+
+val bitwise_overflow_assert : int -> int -> (int -> int -> int) -> unit
+
+val no_32_bit_overflow : int -> int -> (int -> int -> int) -> bool
+
+type rule =
+  Cfg.basic Cfg.instruction DLL.cell ->
+  Cfg.basic Cfg.instruction DLL.cell option
```
[the remark this answers is not in the document]
Comment (xclerc) [reply backend/peephole/peephole_utils.mli:28]: I think `Peephole_rules` will eventually be generated.

File: backend/peephole/peephole_rules.ml [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+          in
+          DLL.delete_curr fst;
+          DLL.delete_curr snd;
+          Some ((prev_at_most go_back_const) new_cell)
+        | _ -> None)
+      | _ -> None
+    else None
+  | _ -> None
+
+let handbuilt_rules cell =
```
[the remark this answers is not in the document]
Comment (xclerc) [reply backend/peephole/peephole_rules.ml:182]: That code would also be generated; we wanted
to avoid `caml_apply`.

File: backend/peephole/peephole_utils.ml [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+[@@@ocaml.warning "+a-29-40-41-42-4"]
+
+module DLL = Flambda_backend_utils.Doubly_linked_list
+
+(* CR-soon gtulba-lecu for gtulba-lecu: make sure that this comparison is
+   correct and sufficent. *)
+let are_equal_regs (reg1 : Reg.t) (reg2 : Reg.t) =
+  Reg.same_loc reg1 reg2 && reg1.typ = reg2.typ
```
[the remark this answers is not in the document]
Comment (xclerc) [reply backend/peephole/peephole_utils.ml:6]: I think `Proc.regs_are_volatile` unconditionally
returns `false` for the two backends in the
repository.

File: backend/.ocamlformat-enable (hunk 1 above)
[the remark this answers is not in the document]
Comment (xclerc) [reply backend/.ocamlformat-enable]: (Already removed.)

File: driver/flambda_backend_args.ml (hunk 8 above)
[the remark this answers is not in the document]
Comment (xclerc) [reply driver/flambda_backend_args.ml]: (Already removed.)

File: backend/csv_data.mli (hunk 5 above)
[the remark this answers is not in the document]
Comment (xclerc) [reply backend/csv_data.mli]: (Already removed.)

File: backend/peephole/peephole_rules.ml [hunk 23]
```diff
@@ -0,0 +1,182 @@
...
+
+   Where <op 1> and <op 2> can be any two binary operators that are associative
+   and const1 and const2 are immediate values. *)
+
+let are_compatible op1 op2 imm1 imm2 =
+  match (op1 : Mach.integer_operation), (op2 : Mach.integer_operation) with
+  (* Folding two bitwise operations such as (AND, OR, XOR) should never produce
+     an overflow. *)
+  | Mach.Iand, Mach.Iand ->
+    bitwise_overflow_assert imm1 imm2 ( land );
```
[the remark this answers is not in the document]
Comment (xclerc) [reply backend/peephole/peephole_rules.ml:53]: (As you pointed out in another comment,
that should really not happen as `imm1`
and `imm2` should arrive to that point
within bounds.)

File: backend/peephole/peephole_rules.ml [hunk 24]
```diff
@@ -0,0 +1,182 @@
...
+    if Misc.no_overflow_add imm1 imm2 && imm1 + imm2 <= Sys.int_size
+    then Some (Mach.Ilsr, imm1 + imm2)
+    else None
+  | Iasr, Iasr ->
+    if Misc.no_overflow_add imm1 imm2 && imm1 + imm2 <= Sys.int_size
+    then Some (Mach.Iasr, imm1 + imm2)
+    else None
+  (* for the amd64 instruction set the `ADD` `SUB` `MUL` opperations take at
+     most an imm32 as the second argument, so we need to check for overflows on
+     32-bit signed ints. *)
```
[the remark this answers is not in the document]
Comment (xclerc) [reply backend/peephole/peephole_rules.ml:86]: I would rather promote the comment to a CR
for now, since we only really support amd64.

File: backend/peephole/peephole_rules.ml [hunk 25]
```diff
@@ -0,0 +1,182 @@
...
+
+let fold_intop_imm (cell : Cfg.basic Cfg.instruction DLL.cell) =
+  match get_cells cell 2 with
+  | [fst; snd] ->
+    let fst_val = DLL.value fst in
+    let snd_val = DLL.value snd in
+    (* The following check does the following: 1. Ensures that both instructions
+       use the same source register; 2. Ensures that both instructions output
+       the result to the source register, this is redundant for amd64 since
+       there are no instructions that invalidate this condition. *)
```
[the remark this answers is not in the document]
Comment (xclerc) [reply backend/peephole/peephole_rules.ml:151]: That makes sense in principle, but I don't
think that should happen in this pull
request.

[the remark this answers is not in the document]
Comment (lthls) [reply backend/peephole/peephole_rules.ml:162]: If you want to have specialised versions of the module, please consider an approach that doesn't use objects, like the one I proposed in https://github.com/ocaml/ocaml/pull/11034.

File: backend/peephole/peephole_rules.ml [hunk 26]
```diff
@@ -1,7 +1,8 @@
 (* CR-someday: see whether the `-4` can be dropped. *)
 [@@@ocaml.warning "+a-29-40-41-42-4"]
 
-open! Peephole_utils
+module DLL = Peephole_utils.DLL
```
Comment (gretay-js) [inline backend/peephole/peephole_rules.ml:4]:
```suggestion
module DLL = Flambda_backend_utils.Doubly_linked_list
```

**Review by gretay-js — APPROVED** [review]

File: backend/peephole/peephole_rules.ml [hunk 27]
```diff
@@ -0,0 +1,182 @@
...
+    match fst_val.desc with
+    | Op (Move | Spill | Reload) -> (
+      let fst_src, fst_dst = fst_val.arg.(0), fst_val.res.(0) in
+      match snd_val.desc with
+      | Op (Move | Spill | Reload) ->
+        let snd_src, snd_dst = snd_val.arg.(0), snd_val.res.(0) in
+        if are_equal_regs fst_src snd_dst && are_equal_regs fst_dst snd_src
+        then (
+          DLL.delete_curr snd;
+          Some (prev_at_most go_back_const fst))
```
[the remark this answers is not in the document]
Comment (gretay-js) [reply backend/peephole/peephole_rules.ml]: Discussed offline, will be addressed in a subsequent PR where Peephole_rules is autogenerated.
